### ISSUE TITLE
Sometimes broken TextMate coloring in header files

### ISSUE THREAD

**Issue by marcdumais-work** [eclipse-theia/theia-cpp-extensions#100]
**Description**
~~This is a slightly tricky issue to reproduce - there are many moving pieces. TL;DR: when the `clangd` language server is unable to be started, it can somehow result in some header files not being colored correctly using the TextMate grammar.~~

Coloring of comments in header files is sometimes broken. It seems to happen with almost any `.h` file that has multi-line comments. 

**How to reproduce:**
Use the latest version of this repo's example Electron application (browser version untested). The `vscode.cpp` built-in extension used should be `v1.43.2`.

I have noticed that coloring seems to be cached. It seems that resetting the workbench layout clears the cache.

~~0) You must have `clangd` installed on your machine and available in the PATH~~
1) start the Electron example app from this repo, using the test workspace that's part of this repo (`examples/cpp-debug-workspace`)
2) Use the sample `.h` file below - save it in your test workspace as `test.h`. 
3) confirm the coloring of the file is ok (see example below)
3a) if not, reset the workbench layout and open the file again. Confirm that the coloring is ok
~~4) Set preference `vscode.clangdExecutable` to a bogus executable name that will not be found. This will result in a failure to start the LS.~~
~~5) reset the workbench layout.~~
5.99) leaving `test.h` open, close the app and start it again (or reload browser tab)
6) ~~Try opening `test.h` again.~~ Confirm that the coloring is broken (see example below)
~~6a) If the coloring is still ok, try closing the example app while leaving `test.h` open, then restart the app, so that it will restart with that file opened and in focus. Confirm that the coloring is broken~~

~~For a while I had the theory that perhaps the TextMate coloring is always broken for `test.h`, but that `semantic highlighting` info provided by `clangd` helps correct the coloring immediately, hiding the issue. However it does not seem to be the case, since I am able to get good coloring without clangd present, so long as `test.h` is not the first C/C++ file opened:~~
__update: not related to LS at all, see comments below__

~~0) set preference `vscode.clangdExecutable` to a bogus executable name~~
~~1) reset the workspace layout~~
~~2) open file `a.cpp` from the example workspace.~~
~~3) then open file `test.h`~~
~~4) confirm that the coloring is ok, even without clangd's help~~

Conclusion: weird race condition that happens when ~~`clangd` cannot be started and~~ we have TextMate coloring being applied ~~at the same time, for some specific `.h` file content (not generally)? I suspect @theia/cpp does not handle well the case where the LS cannot be started.~~

**Expected behavior**
coloring should not be broken

**Environment**
 - _OS_: Ubuntu 18.04
 - _Browser_: Electron
 - _Theia Version_: 1.1.0-next.743836ed

**Additional information**

Here is a `.h` file that can be used to reproduce the issue:
``` cpp
/* > Description *********************************************************************/
/**
 * sdads das asd asd asdasdasdasedasdsdas  dsa asd asd sd asdasd 
 * as dasd sd ad sad sad asd sad sad asd asdasdasdasdasdasdsadasddas
 *
 *****************************************************************************/
 /**
  * @file asdasasdasdasdasdasd.h
  * @brief
  * @see
  *
 *****************************************************************************/

/*>a************/
#ifndef asdasdasdasdasdasd
#define asdasdasdasdasdasd
#endif`
```

Example of ok coloring:
![image](https://user-images.githubusercontent.com/25749063/77953896-cf33d780-729b-11ea-8a14-25f8791b4459.png)

Example of buggy coloring:
![image](https://user-images.githubusercontent.com/25749063/77953851-bb887100-729b-11ea-9430-cc1cc1cd3500.png)

**Comment by marcdumais-work** [comment]
cc: @marechal-p

**Comment by marechal-p** [comment]
On my side, the coloration fails whenever I reload the page. Like you said, doing `Reset Workbench Layout` fixes it, but it breaks as soon as I do `Ctrl +R`. Didn't even have to play with clangd's path.

**Comment by marcdumais-work** [comment]
> On my side, the coloration fails whenever I reload the page. Like you said, doing `Reset Workbench Layout` fixes it, but it breaks as soon as I do `Ctrl +R`. Didn't even have to play with clangd's path.

Yeah, I just confirmed similar - in the end I removed `@theia/cpp` from the electron example and it still breaks after a reload, if it's open in the editor.

I noticed it seems to fall back on its feet it I clear the local storage. So I now suspect a bug there - in the save/restore of the `test.h` editor.

**Comment by marechal-p** [comment]
It looks to be some kind of race condition between the time when the editor is opened, and when grammars are actually loaded/used.

**Comment by marcdumais-work** [comment]
In the local storage, if I clear the value for key `theia:file://<workspace>:editor-navigation-contribution`, close the `test.h` editor and restart the app, it falls back on its feet.

If I do the same but keep the `test.h` editor open, it restarts broken (the value I cleared in local storage is re-saved on exit, including the entry for the `test.h` editor.

**Comment by marcdumais-work** [comment]
> It looks to be some kind of race condition between the time when the editor is opened, and when grammars are actually loaded/used.

Sounds like a good theory. Can you have a troubleshoot?

**Comment by vince-fugnitto** [comment]
Does the error only happen in comments?

**Comment by marcdumais-work** [comment]
> Does the error only happen in comments?

Yes, as far as I know.

**Comment by vince-fugnitto** [comment]
> > Does the error only happen in comments?
> 
> Yes, as far as I know.

I mentioned it because it looks like an bug we get for other languages as well (ex: `yaml`):
- https://github.com/gitpod-io/gitpod/issues/1326

**Comment by marcdumais-work** [comment]
It could be related.

**Comment by marechal-p** [comment]
![image](https://user-images.githubusercontent.com/14182238/78723743-050b3880-78fa-11ea-895c-31570b66a99c.png)

So this is indeed a race condition. When we reload a window when a cpp editor was opened, somehow the editor tries to activate the `cpp` language when we did not finish contributing all grammars.

@akosyakov I have no idea what to do here. Do you have guidance? I am having a hard time figuring out what happens in this area of the code, I think layout, editors, and plugins are related but I don't understand how.

**Comment by marcdumais-work** [comment]
Closing in favor of https://github.com/eclipse-theia/theia/pull/7544

TL;DR the issue is generic - we just happened to notice it with a `cpp` example
